viewer/event_db.cpp

```diff
@@ -1,38 +1,122 @@
+    {
+        _events[klass_id] = std::vector<parser::Event>(1, std::move(event));
+    }
+    else
+    {
+        std::vector<parser::Event>::reverse_iterator it;
+        for (it = _events[klass_id].rbegin(); it != _events[klass_id].rend(); ++it)
+        {   
+            if ((*it).get_timestamp() < event.get_timestamp())
+            {
+                break;
+            } 
+        }
+        _events[klass_id].insert(it.base(), std::move(event));
+    }
+}
+
+void inplace_merge_events(std::vector<EventRef>& response, std::vector<int>& length, int l, int r)
+{
+    if (l == r)
+    {
+        return;
+    }
+
+    int offset_begin = l ? length[l - 1] : 0;
+    int offset_end = length[r];
+    int offset_middle = (offset_begin + offset_end) >> 1;
+
+    if (l + 1 == r)
+    {
+        std::inplace_merge(response.begin() + offset_begin, response.begin() + offset_middle, response.begin() + offset_end,
+                [](const EventRef& e1, const EventRef& e2) {
+                return e1.get().get_timestamp() < e2.get().get_timestamp();
+                });
+        return;
+    }
+
+    int m = (l + r) >> 1;
+    inplace_merge_events(response, length, l, m);
+    inplace_merge_events(response, length, m + 1, r);
+
+    std::inplace_merge(response.begin() + offset_begin, response.begin() + offset_middle, response.begin() + offset_end,
+            [](const EventRef& e1, const EventRef& e2) {
+            return e1.get().get_timestamp() < e2.get().get_timestamp(); 
+            });
 }
 
 std::vector<EventRef> EventDB::get_data(HT_TimestampNs start_ts, HT_TimestampNs stop_ts, const Query& query)
 {
-    std::vector<EventRef> events;
+    std::vector<EventRef> response;
+
+    // Dummy events for calling lower_bound
+    DummyEvent dummy_start_event(start_ts);
+    DummyEvent dummy_stop_event(stop_ts);
 
-    for (const auto& event : _events)
+    if (query.klass_id == (HT_EventKlassId)-1)
     {
-        auto ts = event.get_timestamp();
-        bool get_id_ok = query.klass_id == (HT_EventKlassId)-1 || event.get_klass()->get_id() == query.klass_id;
-        bool ts_ok = ts >= start_ts && ts <= stop_ts;
+        // Query all the event klasses
+        std::vector<int> lengths(_events.size(), 0);
+        int klass_index = -1;
+        for (const auto& events : _events)
+        {
+            const auto first_event = std::lower_bound(events.second.begin(), events.second.end(), dummy_start_event,
+                    [] (const parser::Event& e1, const parser::Event& e2) {
+                    return e1.get_timestamp() < e2.get_timestamp();
+                    });
+            const auto last_event = std::lower_bound(events.second.begin(), events.second.end(), dummy_stop_event,
+                    [] (const parser::Event& e1, const parser::Event& e2) {
+                    return e1.get_timestamp() <= e2.get_timestamp();
+                    });
+            ++klass_index;
+            for (auto it = first_event; it != last_event; ++it)
```

Comment:
looks like you have the same for loop in the `else` branch; could you move this code to separated function, instead of duplicating it?